design-proposals/klip-24-key-column-semantics-in-queries.md

```diff
@@ -0,0 +1,506 @@
+-- resulting schema: ID INT KEY, I1_V0 INT, I1_V1 INT, I2_V0 INT, I2_V1 INT
+```
+
+A more correct implementation might store both sides of the join criteria in the key for all join
+types. However, such an approach would require support for multiple key columns and extensive
+changes in Streams. See [rejected alternatives](#rejected_alternatives) for more info.
+
+### Removal of non-standard Aliasing
+
+The [allow any key name](https://github.com/confluentinc/ksql/issues/3536) feature has introduced
+the ability to provide an alias within a `PARTITION BY`, `GROUP BY` and `JOIN` clause. These are
+non-standard, but required in the current model to allow users to name the key column in the schema.
+For example:
+
+```sql
+-- without alias:
+CREATE STREAM OUTPUT AS SELECT ID FROM INPUT PARTITION BY V0 - V1;
+-- resulting schema: KSQL_COL_0 INT KEY, ID;
+-- note the system generated column name.
+
+-- with alias
+CREATE STREAM OUTPUT AS SELECT ID FROM INPUT PARTITION BY V0 - V1 AS NEW_KEY;
+-- resulting schema: NEW_KEY INT KEY, ID;
+```
+
+However, the same functionality can be achieved using standard sql if the key column is required
+in the projection, for example:
+
+```sql
+CREATE STREAM OUTPUT AS SELECT V0 - V1 AS NEW_KEY, ID FROM INPUT PARTITION BY V0 - V1;
+-- resulting schema: NEW_KEY INT KEY, ID;
+```
+
+### Any key name
+
+Persistent queries failing if the projection contained the key column existed before the work to
+[allow any key name](https://github.com/confluentinc/ksql/issues/3536), though this work also
+exacerbated the situation, as the key column is no longer always called `ROWKEY`. Instead, users can
+pick any name for the key column, and the key column takes on the name of any `PARTITION BY`,
+`GROUP BY` or `JOIN ON` on a single column. This causes seemingly correct and common patterns to
+fail with duplicate column errors:
+
+```sql
+-- Before 'any key name':
+CREATE STREAM INPUT (ROWKEY INT KEY, V0 INT, V1 INT) WITH (...);
+CREATE TABLE OUTPUT AS SELECT V0, COUNT(*) AS COUNT FROM INPUT GROUP BY V0;
+-- resulting schema: ROwKEY INT KEY, V0 INT, COUNT BIGINT
+-- Note: ROWKEY & V0 store the same data, which isn't ideal.
+
+-- vs --
+
+-- After 'any key name':
+CREATE STREAM INPUT (ID INT KEY, V0 INT, V1 INT) WITH (...);
+CREATE TABLE OUTPUT AS SELECT V0, COUNT(*) AS COUNT FROM INPUT GROUP BY V0;
+-- fails with duplicate column error on 'V0'.
+
+-- Note: the equivalent query would have failed before the any-key-name work:
+CREATE TABLE OUTPUT AS SELECT ROWKEY, COUNT(*) AS COUNT FROM INPUT GROUP BY V0;
+-- fails with duplicate column error on 'ROWKEY'.
+```
+
+As you can see from above, the common pattern of selecting the group by key and the aggregate fails.
+To 'fix' the query the user must remove `V0` from the projection - which is counter intuitive, as
+the user wishes this column in the result.
+
+We propose that this query should work, without any modification, and without storing duplicate
```

Comment:
I guess, only the first query should work. After this KIP is merged, the second query should have a column called `ROWKEY` and thus should fail?

Reply:
clarified.